project_euler/problem_107/sol1.c

```diff
@@ -0,0 +1,188 @@
+ * @param i the first vertex
+ * @param j the second vertex
+ * @param uf the union find i and j lie within
+ * @returns true if i and j are in the same set
+ * @returns false otherwise
+ */
+int in_same_set(int i, int j, struct ufds *uf) {
+  return find_set(i, uf) == find_set(j, uf);
+}
+
+/** Bring together two sets in a union find
+ * @param i the first representative item of one set
+ * @param j the second representative item of another set
+ * @param uf the union find they lie within
+ */
+void union_set(int i, int j, struct ufds *uf) {
+  if (!in_same_set(i, j, uf)) {
+    int ri = find_set(i, uf), rj = find_set(j, uf);
+    if (uf->rank[ri] < uf->rank[rj]) {
+      uf->parent[ri] = rj;
+      uf->size[rj]++;
+    } else {
+      uf->parent[rj] = ri;
+      if (uf->rank[ri] == uf->rank[rj]) {
+        uf->rank[rj]++;
+        uf->size[rj]++;
+      }
+    }
+    uf->num_sets--;
+  }
+}
+
+/*
+ * The problem specifies an adjacency matrix, which can be used with Prim's
+ * algorithm, but this implementation uses kruskal's, so we convert the input
+ * into an edge list represented by an array of edges
+ */
+void generate_edge_list(char *str, struct edge *elist, int *total_weight) {
+  char *tmp;
+  int n;
+  tmp = strtok(str, ",");
+  if (!tmp) {
+    fprintf(stderr, "input is invalid!\n");
+    exit(EXIT_FAILURE);
+  }
+  for (int i = 0; i < NUM_VERTICES; i++) {
+    for (int j = (i == 0) ? 1 : 0; j < NUM_VERTICES; j++) {
+      struct edge *entry = &elist[i * NUM_VERTICES + j];
+      tmp = strtok(NULL, ",");
+      // ignore redundant data symmetric about the line i == j in the matrix
+      if (j < i || strcmp(tmp, "-") == 0) {
+        continue;
+      } else {
+        n = atoi(tmp);
+        entry->u = i;
+        entry->v = j;
+        entry->w = n;
+        *total_weight += n;
+      }
+    }
+  }
+}
+
+/* Compare two edges by their weight */
+int my_comp(const void *a, const void *b) {
+  return ((struct edge *)a)->w - ((struct edge *)b)->w;
+}
+
+/** An implementation of Kruskal's algorithm
+ * (https://en.wikipedia.org/wiki/Kruskal%27s_algorithm)
+ * @param E the number of edges
+ * @param elist the edge list
+ * @param uf pointer to union find for keeping track of trees
+ */
+int kruskals(int E, struct edge *elist, struct ufds *uf) {
+  int cost = 0;
+  // This is the main cost of kruskal's - O(ElogE) = O(Elog(V^2)) = O(ElogV)
+  qsort(elist, NUM_VERTICES * NUM_VERTICES, sizeof(struct edge), my_comp);
+  for (int i = 0; i < E; i++) {
+    struct edge *e = &elist[i];
+    if (!in_same_set(e->u, e->v, uf)) {
+      cost += e->w;
+      union_set(e->u, e->v, uf);
+    }
+    if (uf->num_sets == 1) {
+      return cost;
+    }
+  }
+  return cost;
+}
+
+/*
+ * We parse input and generate an edge list
+ * we then pass that edge list into kruskals to
+ * get the cost of a MST of the network.
+ * We then subtract that from the total cost to
+ * get the savings and the solution to the problem.
+ */
+
+/** Main function */
```

Comment:
```suggestion
/** Main function 
 * @return 0
 */
```

Reply:
Fixed in 67e24f42cf2c3a11b780ff3b2af496002cc2ae71